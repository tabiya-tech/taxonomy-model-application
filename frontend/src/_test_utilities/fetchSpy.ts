export function setupFetchSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });
  return jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
}
