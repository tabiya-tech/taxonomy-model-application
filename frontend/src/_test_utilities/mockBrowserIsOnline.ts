import { fireEvent } from "@testing-library/react";

export function mockBrowserIsOnLine(isOnline: boolean) {
  if (window.navigator.onLine !== isOnline) {
    jest.spyOn(window.navigator, "onLine", "get").mockReturnValue(isOnline);
    if (isOnline) {
      fireEvent.online(window);
    } else {
      fireEvent.offline(window);
    }
  }
}

export function unmockBrowserIsOnLine() {
  jest.spyOn(window.navigator, "onLine", "get").mockRestore();
}
