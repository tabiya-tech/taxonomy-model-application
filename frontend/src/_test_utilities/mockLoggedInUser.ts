import { AuthContextValue, TabiyaUser, TabiyaUserRole, authContextDefaultValue } from "src/auth/AuthProvider";

export function mockLoggedInUser(mockedValues: Partial<AuthContextValue>){
  let user: TabiyaUser = mockedValues.user || { username: "foo", roles: [] };
  jest.spyOn(require("src/auth/hooks/useAuthUser"), "useAuthUser").mockReturnValue({
    ...authContextDefaultValue,
    user: user,
    setUser: (user: TabiyaUser) => null,
    hasRole: (role: TabiyaUserRole) => user.roles.includes(role),
    updateUserByAccessToken: () => {},
    ...mockedValues
  });
}
