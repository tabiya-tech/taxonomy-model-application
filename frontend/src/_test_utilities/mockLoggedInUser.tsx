import {
  AuthContextValue,
  TabiyaUser,
  authContextDefaultValue,
  AuthContext,
  AuthProviderProps,
} from "src/auth/AuthProvider";
import AuthAPISpecs from "api-specifications/auth";

export function mockLoggedInUser(mockedValues: Partial<AuthContextValue>) {
  let user: TabiyaUser = mockedValues.user || TestUsers.Anonymous;

  const value = {
    ...authContextDefaultValue,
    user: user,
    setUser: (user: TabiyaUser) => null,
    hasRole: (role: AuthAPISpecs.Enums.TabiyaRoles) => user.roles.includes(role),
    updateUserByAccessToken: () => {},
    ...mockedValues,
  };

  // @ts-ignore
  jest.spyOn(require("src/auth/AuthProvider"), "AuthProvider").mockImplementation((props: AuthProviderProps) => {
    return <AuthContext.Provider value={value}>{props.children}</AuthContext.Provider>;
  });
}

export const TestUsers = {
  ModelManager: {
    username: "ModelManagerUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.MODEL_MANAGER],
  },
  RegisteredUser: {
    username: "RegisteredUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.REGISTERED_USER],
  },
  Anonymous: {
    username: "AnonymousUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.ANONYMOUS],
  },
};
