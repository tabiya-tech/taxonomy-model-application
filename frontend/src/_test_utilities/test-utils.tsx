// Based on https://testing-library.com/docs/react-testing-library/setup/

import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/providers";

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
