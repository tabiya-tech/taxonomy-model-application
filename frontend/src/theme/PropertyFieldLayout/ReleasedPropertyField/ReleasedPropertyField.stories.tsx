import ReleasedPropertyField from "./ReleasedPropertyField";
import { Meta, StoryObj } from "@storybook/react";

const meta: Meta = {
  title: "components/ReleasedPropertyField",
  component: ReleasedPropertyField,
  tags: ["autodocs"],
  argTypes: {},
};

export default meta;

type Story = StoryObj<typeof ReleasedPropertyField>;

export const ShownWithReleased: Story = {
  args: {
    released: true,
  },
};

export const ShownWithNotReleased: Story = {
  args: {
    released: false,
  },
};
