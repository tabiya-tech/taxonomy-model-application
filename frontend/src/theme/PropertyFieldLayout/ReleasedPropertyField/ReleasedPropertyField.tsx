import React from "react";
import { Typography } from "@mui/material";
import PropertyFieldLayout from "../PropertyFieldLayout";
import PublishedWithChangesIcon from "@mui/icons-material/PublishedWithChanges";
import UnpublishedOutlinedIcon from "@mui/icons-material/UnpublishedOutlined";
import Box from "@mui/material/Box";

export interface ReleasedPropertyFieldProps {
  released: boolean;
  "data-testid"?: string;
  fieldId: string;
}

const uniqueId = "7e51bcc0-7abc-462c-81e8-a33058fa1b1e";

export const DATA_TEST_ID = {
  RELEASED_PROPERTY_FIELD_STATUS: `released-property-field-status-${uniqueId}`,
  RELEASED_ICON: `released-icon-${uniqueId}`,
  NOT_RELEASED_ICON: `not-released-icon-${uniqueId}`,
  RELEASED_TEXT: `released-text-${uniqueId}`,
  NOT_RELEASED_TEXT: `not-released-text-${uniqueId}`,
};

export const TEXT = {
  RELEASED: "Released",
  NOT_RELEASED: "Not Released",
};

const ReleasedPropertyField = (props: Readonly<ReleasedPropertyFieldProps>) => {
  return (
    <PropertyFieldLayout title="Released Status" fieldId={props.fieldId} data-testid={props["data-testid"]}>
      <Box
        display="flex"
        gap={(theme) => theme.tabiyaSpacing.xs}
        alignItems="center"
        data-testid={DATA_TEST_ID.RELEASED_PROPERTY_FIELD_STATUS}
        aria-labelledby={props.fieldId}
        role="text"
      >
        {props.released ? (
          <>
            <PublishedWithChangesIcon
              sx={{ color: (theme) => theme.typography.body1.color }}
              data-testid={DATA_TEST_ID.RELEASED_ICON}
            />
            <Typography variant="body1" data-testid={DATA_TEST_ID.RELEASED_TEXT}>
              {TEXT.RELEASED}
            </Typography>
          </>
        ) : (
          <>
            <UnpublishedOutlinedIcon
              sx={{ color: (theme) => theme.typography.body1.color }}
              data-testid={DATA_TEST_ID.NOT_RELEASED_ICON}
            />
            <Typography variant="body1" data-testid={DATA_TEST_ID.NOT_RELEASED_TEXT}>
              {TEXT.NOT_RELEASED}
            </Typography>
          </>
        )}
      </Box>
    </PropertyFieldLayout>
  );
};

export default ReleasedPropertyField;
