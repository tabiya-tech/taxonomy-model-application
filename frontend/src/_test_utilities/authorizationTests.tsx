import { render, screen } from "src/_test_utilities/test-utils";
import AuthAPISpecs from "api-specifications/auth";
import { mockLoggedInUser } from "./mockLoggedInUser";

export const ALL_USERS = [
  AuthAPISpecs.Enums.TabiyaRoles.ANONYMOUS,
  AuthAPISpecs.Enums.TabiyaRoles.REGISTERED_USER,
  AuthAPISpecs.Enums.TabiyaRoles.MODEL_MANAGER,
];

type AuthorizationTestsParams = {
  name: string;
  Component: JSX.Element;
  testIds?: string[];
  roles?: AuthAPISpecs.Enums.TabiyaRoles[];
  tests?: Partial<
    Record<
      AuthAPISpecs.Enums.TabiyaRoles,
      {
        visible?: string[];
        hidden?: string[];
        visibleDisabled?: string[];
      }
    >
  >;
};

export const authorizationTests = {
  defaultName: "Test the rendering of the component based on Roles",
  callback: ({ name, Component, testIds = [], roles = ALL_USERS, tests }: AuthorizationTestsParams) => {
    return () => {
      test.each(roles?.map((role) => [role]))(`${name} renders correctly for user with role %s`, (role) => {
        const commonElements = testIds || [];
        const { visible = [], visibleDisabled = [], hidden = [] } = tests?.[role] || {};

        mockLoggedInUser({ user: { username: "foo", roles: [role] } });

        render(Component);

        commonElements.forEach((testId) => expect(screen.getByTestId(testId)).toBeInTheDocument());

        visible?.forEach((testId) => {
          expect(screen.getByTestId(testId)).toBeVisible();
        });

        hidden?.forEach((testId) => {
          expect(screen.queryByTestId(testId)).not.toBeInTheDocument();
        });

        visibleDisabled?.forEach((testId) => {
          const element = screen.getByTestId(testId);
          expect(element).toBeVisible();
          if (element instanceof HTMLInputElement) {
            expect(element).toBeDisabled();
          }
        });
      });
    };
  },
};
