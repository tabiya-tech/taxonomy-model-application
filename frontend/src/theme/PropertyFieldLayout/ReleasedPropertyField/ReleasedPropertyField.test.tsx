// mute the console
import "src/_test_utilities/consoleMock";

import React from "react";
import ReleasedPropertyField, { DATA_TEST_ID, TEXT } from "./ReleasedPropertyField";
import { render, screen, within } from "src/_test_utilities/test-utils";

describe("ReleasedPropertyField", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    (console.error as jest.Mock).mockClear();
    (console.warn as jest.Mock).mockClear();
  });

  test.each([
    [
      "Released",
      true,
      {
        expectedIcon: DATA_TEST_ID.RELEASED_ICON,
        expectedText: TEXT.RELEASED,
        expectedTextTestId: DATA_TEST_ID.RELEASED_TEXT,
      },
    ],
    [
      "Not Released",
      false,
      {
        expectedIcon: DATA_TEST_ID.NOT_RELEASED_ICON,
        expectedText: TEXT.NOT_RELEASED,
        expectedTextTestId: DATA_TEST_ID.NOT_RELEASED_TEXT,
      },
    ],
  ])("should render with provided '%s' status", (_description, givenReleaseStatus, expectedElements) => {
    // GIVEN a released status
    // AND a valid fieldId
    const givenFieldId = "field-id";
    // AND a data-testid
    const givenDataTestId = "released-details";

    // WHEN the component is rendered
    render(
      <ReleasedPropertyField released={givenReleaseStatus} data-testid={givenDataTestId} fieldId={givenFieldId} />
    );

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
    // AND the component to be displayed
    const expectedReleasedPropertyComponent = screen.getByTestId(givenDataTestId);
    expect(expectedReleasedPropertyComponent).toBeInTheDocument();
    // AND the tile to be shown
    const expectedTitle = screen.getByText("Released Status");
    expect(expectedTitle).toBeInTheDocument();
    // AND the released property field status to be shown
    const expectedReleasedPropertyFieldStatus = screen.getByTestId(DATA_TEST_ID.RELEASED_PROPERTY_FIELD_STATUS);
    expect(expectedReleasedPropertyFieldStatus).toBeInTheDocument();
    // AND the Typography component to be shown
    const expectedTypographyComponent = screen.getByTestId(expectedElements.expectedTextTestId);
    expect(expectedTypographyComponent).toBeInTheDocument();
    // AND the released status text to be shown wrapped in the Typography component
    const expectedReleasedStatusText = within(expectedTypographyComponent).getByText(expectedElements.expectedText);
    expect(expectedReleasedStatusText).toBeInTheDocument();
    // AND the icon to be shown
    const expectedIcon = screen.getByTestId(expectedElements.expectedIcon);
    expect(expectedIcon).toBeInTheDocument();
    // AND the component to match the snapshot
    expect(expectedReleasedPropertyComponent).toMatchSnapshot();
  });
});
