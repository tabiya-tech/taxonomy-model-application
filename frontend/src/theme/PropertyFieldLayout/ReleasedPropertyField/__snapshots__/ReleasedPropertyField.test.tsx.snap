// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReleasedPropertyField should render with provided 'Not Released' status 1`] = `
<div
  class="MuiBox-root css-1ofqig9"
  data-testid="released-details"
>
  <h6
    class="MuiTypography-root MuiTypography-h6 css-1h8vygf-MuiTypography-root"
    data-testid="item-title-5abb0d55-4073-4f5b-8432-ee8c96aa4df2"
    id="field-id"
    tabindex="0"
  >
    Released Status
  </h6>
  <div
    aria-labelledby="field-id"
    class="MuiBox-root css-1vvwjbg"
  >
    <div
      aria-labelledby="field-id"
      class="MuiBox-root css-i8j0ju"
      data-testid="released-property-field-status-7e51bcc0-7abc-462c-81e8-a33058fa1b1e"
      role="text"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-il7e6s-MuiSvgIcon-root"
        data-testid="not-released-icon-7e51bcc0-7abc-462c-81e8-a33058fa1b1e"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7.94 5.12 6.49 3.66C8.07 2.61 9.96 2 12 2c5.52 0 10 4.48 10 10 0 2.04-.61 3.93-1.66 5.51l-1.46-1.46C19.59 14.86 20 13.48 20 12c0-4.41-3.59-8-8-8-1.48 0-2.86.41-4.06 1.12zm9.72 4.41-1.41-1.41-2.65 2.65 1.41 1.41 2.65-2.65zm2.12 13.08-2.27-2.27C15.93 21.39 14.04 22 12 22 6.48 22 2 17.52 2 12c0-2.04.61-3.93 1.66-5.51L1.39 4.22 2.8 2.81l18.38 18.38-1.4 1.42zm-3.72-3.73L12.18 15l-1.59 1.59-4.24-4.24 1.41-1.41 2.83 2.83.18-.18-5.65-5.65C4.41 9.14 4 10.52 4 12c0 4.41 3.59 8 8 8 1.48 0 2.86-.41 4.06-1.12z"
        />
      </svg>
      <p
        class="MuiTypography-root MuiTypography-body1 css-3ahvbo-MuiTypography-root"
        data-testid="not-released-text-7e51bcc0-7abc-462c-81e8-a33058fa1b1e"
      >
        Not Released
      </p>
    </div>
  </div>
</div>
`;

exports[`ReleasedPropertyField should render with provided 'Released' status 1`] = `
<div
  class="MuiBox-root css-1ofqig9"
  data-testid="released-details"
>
  <h6
    class="MuiTypography-root MuiTypography-h6 css-1h8vygf-MuiTypography-root"
    data-testid="item-title-5abb0d55-4073-4f5b-8432-ee8c96aa4df2"
    id="field-id"
    tabindex="0"
  >
    Released Status
  </h6>
  <div
    aria-labelledby="field-id"
    class="MuiBox-root css-1vvwjbg"
  >
    <div
      aria-labelledby="field-id"
      class="MuiBox-root css-i8j0ju"
      data-testid="released-property-field-status-7e51bcc0-7abc-462c-81e8-a33058fa1b1e"
      role="text"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-il7e6s-MuiSvgIcon-root"
        data-testid="released-icon-7e51bcc0-7abc-462c-81e8-a33058fa1b1e"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="m17.66 9.53-7.07 7.07-4.24-4.24 1.41-1.41 2.83 2.83 5.66-5.66 1.41 1.41zM4 12c0-2.33 1.02-4.42 2.62-5.88L9 8.5v-6H3l2.2 2.2C3.24 6.52 2 9.11 2 12c0 5.19 3.95 9.45 9 9.95v-2.02c-3.94-.49-7-3.86-7-7.93zm18 0c0-5.19-3.95-9.45-9-9.95v2.02c3.94.49 7 3.86 7 7.93 0 2.33-1.02 4.42-2.62 5.88L15 15.5v6h6l-2.2-2.2c1.96-1.82 3.2-4.41 3.2-7.3z"
        />
      </svg>
      <p
        class="MuiTypography-root MuiTypography-body1 css-3ahvbo-MuiTypography-root"
        data-testid="released-text-7e51bcc0-7abc-462c-81e8-a33058fa1b1e"
      >
        Released
      </p>
    </div>
  </div>
</div>
`;
